src/client/Microsoft.Identity.Client/ManagedIdentity/ManagedIdentityClient.cs

```diff
@@ -0,0 +1,53 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using Microsoft.Identity.Client.Extensibility;
+using System.Threading.Tasks;
+using System.Threading;
+using System.Linq;
+using Microsoft.Identity.Client.Http;
+using Microsoft.Identity.Client.Internal;
+using Microsoft.IdentityModel.Abstractions;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Utils;
+
+namespace Microsoft.Identity.Client.ManagedIdentity
+{
+    /// <summary>
+    /// Class to initialize a managed identity and identify the service.
+    /// Original source of code: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs
+    /// </summary>
+    internal class ManagedIdentityClient
+    {
+        internal const string MsiUnavailableError =
+            "Authentication with managed identity is unavailable. No managed identity endpoint found.";
+
+        private readonly Lazy<ManagedIdentitySource> _identitySource;
```

Comment:
If the lambda inside the Lazy throws expcetion, the next time you call `Value`, it will execute the lambda again (assuming the exception is caught of course).